### PR TITLE
[Translation] [Portuguese Brazil]: methods article

### PR DESCRIPTION
**Pull request by BFostek**
Portuguese translation of  those articles:

- String Algorithm 
- Generics type
- Struct types

### CODE REVIEW

File: _content/tour/por/generics-type-constraints.article [hunk 1]
```diff
@@ -1,68 +1,66 @@
...
     }
 
-Now the compiler can validate that the set of types is compliant with the operations
-the function needs to perform against values of those types. When the interface is
-using the built-in types, the interfaces are reusable across packages. When the list
-of types represent user-defined types from the package, You must remember these
-generic functions are bound to the packages types and nothing more.
+Agora o compilador pode validar que o conjunto de tipos é compatível com as operações
+que a função precisa realizar com valores daqueles tipos. Quando a interface é
+usando os tipos embutidos, as interfaces são reutilizáveis entre pacotes. Quando a lista
```
Comment (rschio) [inline _content/tour/por/generics-type-constraints.article:55]: I see "embutido" as a better translation of embedded than built-in. When translating some articles I had difficulty to translate built-in too, I thought in "primitivo ou básico", but none of them fit every situation so I didn't translate.
```suggestion
que a função precisa realizar com valores daqueles tipos. Quando a interface está
usando os tipos built-in, as interfaces são reutilizáveis entre pacotes. Quando a lista
```

File: _content/tour/por/generics-type-constraints.article [hunk 2]
```diff
@@ -74,14 +72,13 @@ One idea is to have pre-declared identifiers for common operation constraints.
         return -1
     }
 
-The comparable constraint is declared by the language and applies a constraint that
-types must be capable of being used in a comparison statement. In this example,
-both v and find are variables of type T and are being compared. There is an idea
-that a package in the standard library could provide a common set of constraints as
-well.
+A restrição comparável é declarada pela linguagem e aplica uma restrição que
```
Comment (rschio) [inline _content/tour/por/generics-type-constraints.article:75]: Comparable is a name.

```suggestion
A restrição comparable é declarada pela linguagem e aplica uma restrição que
```

File: _content/tour/por/generics-type-constraints.article [hunk 3]
```diff
@@ -98,26 +95,26 @@ method set of behavior.
...
+person e food.
 
-When you look at the match function, there isn’t an obvious need to restrict the
-function to just the person and food types. If this is the case, the match function
-should be a traditional polymorphic function, not a generic function. If there was a
-good reason, a generic function can be used to apply this type of constraint.
+Quando você olha para a função match, não há uma necessidade óbvia de restringir a
+função apenas aos tipos person e food. Se este for o caso, a função match
+deveria ser uma função polimórfica tradicional, não uma função genérica. Se houvesse um
+bom motivo, uma função genérica pode ser usada para aplicar este tipo de restrição.
```
Comment (rschio) [inline _content/tour/por/generics-type-constraints.article:105]:
```suggestion
deveria ser uma função polimórfica tradicional, não uma função genérica. Se houvesse um
bom motivo, uma função genérica poderia ser usada para aplicar este tipo de restrição.
```

File: _content/tour/por/algorithms-strings.article [hunk 4]
```diff
@@ -1,67 +1,66 @@
...
+- Precisa de Assistência Financeira? Use Nosso [[https://www.ardanlabs.com/scholarship/][Formulário de Bolsa de estudos]]
 
-The sample program implements a check to see if a string is a
-permutation or not.
+O programa exemplo implementa uma verificação para ver se uma string é ou não uma permutação.
 
-- See more at [[https://en.wikipedia.org/wiki/Permutation][https://en.wikipedia.org/wiki/Permutation]]
 
-*Diagram*
+- Veja mais em [[https://pt.wikipedia.org/wiki/Permutação][https://pt.wikipedia.org/wiki/Permutação]]
```
Comment (rschio) [inline _content/tour/por/algorithms-strings.article:12]: I don't think we should change the link.

File: _content/tour/por/algorithms-strings.article [hunk 5]
```diff
@@ -1,67 +1,66 @@
...
-* Palindrome
+* Palíndromo
 
-The sample program implements a check to see if a string is a
-palindrome or not.
+O programa exemplo implementa uma verificação para ver se a string
+é ou não palíndromo. 
 
-- See more at [[https://en.wikipedia.org/wiki/Palindrome][https://en.wikipedia.org/wiki/Palindrome]]
+- Veja mais em [[https://pt.wikipedia.org/wiki/Palíndromo][https://pt.wikipedia.org/wiki/Palíndromo]]
```
Comment (rschio) [inline _content/tour/por/algorithms-strings.article:47]: I don't think we should change the link.

File: _content/tour/por/algorithms-strings.article [hunk 6]
```diff
@@ -1,67 +1,66 @@
...
 
-*Diagram*
+*Diagrama*
 
-    A palindrome is a word, number, phrase, or other sequence of symbols that
-    reads the same backwards as forwards.
+    Um palíndromo é uma palavra, numero, frase, ou outra sequência de símbolos que
```
Comment (rschio) [inline _content/tour/por/algorithms-strings.article:51]:
```suggestion
    Um palíndromo é uma palavra, número, frase, ou outra sequência de símbolos que
```

File: _content/tour/por/struct-types.article [hunk 7]
```diff
@@ -83,39 +82,39 @@ using literal construction syntax.
...
     Pi 3.141592
 
-The idea of literal construction is just that, to construct something literally
-without a name.
+A ideia de construção literal é justamente essa, construir algo literalmente sem um nome.
 
-You should use var for zero value and the short variable declaration operator with
-the { } syntax for non-zero value construction.
+Você deve usar var para valor zero e o operador de declaração de variável curta com a sintaxe 
+{ } para construção de valor não-zero.
```
[the remark this answers is not in the document]
Comment (sergiolpf) [reply _content/tour/por/struct-types.article:94]: concordo que o "valor zero" não faz muito sentido. Acho que talvez usarmos o "valor inicial padrão"?

File: _content/tour/por/algorithms-strings.article (hunk 4 above)
Comment (BFostek) [reply _content/tour/por/algorithms-strings.article:12]: What are you guys doing? @sergiolpf @sergiolpf @rschio

Comment (sergiolpf) [reply _content/tour/por/algorithms-strings.article:12]: I am not translating links. Looking at other languages translations they also did not translate the links.

File: _content/tour/por/algorithms-strings.article (hunk 5 above)
Comment (luciormoraes) [reply _content/tour/por/algorithms-strings.article:47]: I have the same thought @BFostek let's keep the links.

**Comment by BFostek** [comment]
I made the suggested commits and pushed the other files

File: _content/tour/por/constants.article [hunk 8]
```diff
@@ -1,123 +1,121 @@
...
+Uma das características mais singulares de Go é a forma como a linguagem implementa constantes.
+As regras para constantes na especificação da linguagem são exclusivas para Go. Elas oferecem
+a flexibilidade que Go precisa para tornar o código que escrevemos legível e intuitivo, enquanto ainda
+mantém a segurança de tipo.
 
-Constants can be typed or untyped. When a constant is untyped, it’s considered to
-be of a kind. Constants of a kind can be implicitly converted by the compiler. This
-all happens at compile time and not at runtime.
+Constantes podem ser tipadas ou não tipadas. Quando uma constante é não tipada, ela é considerada
+de um certo "tipo" (em inglês "kind"). Constantes desse "tipo" podem ser convertidas implicitamente pelo compilador.
```
Comment (rschio) [inline _content/tour/por/constants.article:15]: I think we don't have a good translation for kind, in my opinion we should not translate. It's very hard to distinguish when reading and impossible when speaking.

File: _content/tour/por/constants.article [hunk 9]
```diff
@@ -1,123 +1,121 @@
...
 
     var answer = 3 * 0.333  // float64 = KindFloat(3) * KindFloat(0.333)
 
-Constant arithmetic supports the use of different kinds of constants. Kind Promotion
-is used to handle these different scenarios. All of this happens implicitly. The answer
-variable in this example will be of type float64 and represent 0.999 at a precision
-of 64 bits.
+A aritmética de constantes suporta o uso de diferentes tipos de constantes. A Promoção de Tipo (em inglês "Kind Promotion")
+é usada para lidar com esses diferentes cenários. Tudo isso acontece implicitamente. A variável "answer"
+neste exemplo será do tipo float64 e representará 0.999 com uma precisão de 64 bits.
```
Comment (rschio) [inline _content/tour/por/constants.article:49]: Here for example, it supports arithmetic between different kinds not types.
```go
const a int = 10
const b float64 = 1.1
const c = a + b // this is invalid
const d = 10 + 1.1 // this is valid
```

File: _content/tour/por/generics-basics.article [hunk 10]
```diff
@@ -39,32 +39,29 @@ and not use reflection, you can use the new generics syntax.
...
+para escrever código genérico.
+
+Deve haver uma maneira de dizer ao compilador que você não irá declarar o tipo T explicitamente, 
+mas ele deve ser determinado pelo compilador durante a compilação. 
+A nova sintaxe usa colchetes quadrados para isso. Os colchetes definem uma lista de identificadores 
+de tipo genéricos que representam tipos específicos para a função que precisam ser determinados durante a compilação. 
+É assim que você diz ao compilador que os tipos com esses nomes não serão declarados antes que o programa seja compilado. 
+Esses tipos precisam ser resolvidos durante a compilação.
+
+Nota: YVocê pode ter vários identificadores de tipo definidos dentro dos colchetes,
```
Comment (rschio) [inline _content/tour/por/generics-basics.article:54]:
```suggestion
Nota: Você pode ter vários identificadores de tipo definidos dentro dos colchetes, 
```

File: _content/tour/por/generics-basics.article [hunk 11]
```diff
@@ -39,32 +39,29 @@ and not use reflection, you can use the new generics syntax.
...
-a well defined constraint. The any constraint is predeclared by the compiler and
-states there are no constraints on what type T can be.
+Esta é uma implementação de uma única função de impressão que pode exibir uma slice de qualquer tipo 
+dado usando a nova sintaxe de genéricos. O que é bom nesta sintaxe é que o código dentro da função pode usar a sintaxe e 
+funções integradas que funcionariam com um tipo concreto. Isso não é o caso quando você usa a interface vazia 
+para escrever código genérico.
+
+Deve haver uma maneira de dizer ao compilador que você não irá declarar o tipo T explicitamente, 
+mas ele deve ser determinado pelo compilador durante a compilação. 
+A nova sintaxe usa colchetes quadrados para isso. Os colchetes definem uma lista de identificadores
```
Comment (rschio) [inline _content/tour/por/generics-basics.article:49]:
```suggestion
A nova sintaxe usa colchetes para isso. Os colchetes definem uma lista de identificadores 
```

File: _content/tour/por/generics-basics.article [hunk 12]
```diff
@@ -75,14 +72,76 @@ states there are no constraints on what type T can be.
     floats := []float64{1.7, 2.2, 3.14}
     print[float64](floats)
 
-This is how to make calls to the generic print function where the type information
-for T is explicitly provided at the call site. The syntax emulates the idea that the
-function declaration func name[T any](slice []T) defines two sets of parameters.
-The first set is the type that maps to the corresponding type identifiers, and the
-second is the data that maps to the corresponding input variables.
+** Básicos
```
Comment (rschio) [inline _content/tour/por/generics-basics.article:75]: Here you copied the file again, please check the original to fix this.

**Review by rschio — COMMENTED** [review]
Most of articles have only small issues, but the "generics-basics" and "constants" need some attention.

File: _content/tour/por/generics-basics.article [hunk 13]
```diff
@@ -75,14 +72,76 @@ states there are no constraints on what type T can be.
...
+strings := []string{"A", "B", "C"}
+print[string](strings)
+
+flutuantes := []float64{1.7, 2.2, 3.14}
+print[float64](flutuantes)
+
+É assim que você faz chamadas para a função genérica de print, onde as informações de tipo para T são fornecidas explicitamente no local da chamada. 
+A sintaxe emula a ideia de que a declaração da função func nome[T any](slice []T) define dois conjuntos de parâmetros. 
+O primeiro conjunto é o tipo que corresponde aos identificadores de tipo correspondentes, e o segundo é os dados que correspondem às variáveis de entrada 
+correspondentes.
```
Comment (rschio) [inline _content/tour/por/generics-basics.article:141]:
```suggestion
```

File: _content/tour/por/generics-basics.article [hunk 14]
```diff
@@ -75,14 +72,76 @@ states there are no constraints on what type T can be.
     floats := []float64{1.7, 2.2, 3.14}
     print[float64](floats)
 
-This is how to make calls to the generic print function where the type information
-for T is explicitly provided at the call site. The syntax emulates the idea that the
-function declaration func name[T any](slice []T) defines two sets of parameters.
-The first set is the type that maps to the corresponding type identifiers, and the
-second is the data that maps to the corresponding input variables.
```
Comment (rschio) [inline _content/tour/por/generics-basics.article:82]: Translate this

File: _content/tour/por/constants.article [hunk 15]
```diff
@@ -1,123 +1,121 @@
...
-This doesn't work because the number 1000 is too large to store in an uint8.
+Isso não funciona porque o número 1000 é grande demais para ser armazenado em um uint8
 
     var answer = 3 * 0.333  // float64 = KindFloat(3) * KindFloat(0.333)
 
-Constant arithmetic supports the use of different kinds of constants. Kind Promotion
-is used to handle these different scenarios. All of this happens implicitly. The answer
-variable in this example will be of type float64 and represent 0.999 at a precision
-of 64 bits.
+A aritmética de constantes suporta o uso de diferentes tipos de constantes. A Promoção de Tipo (em inglês "Kind Promotion")
```
Comment (rschio) [inline _content/tour/por/constants.article:47]:
```suggestion
A aritmética de constantes suporta o uso de diferentes "kinds" de constantes. A Promoção de Tipo (em inglês "Kind Promotion")
```

File: _content/tour/por/constants.article [hunk 16]
```diff
@@ -1,123 +1,121 @@
...
     const third = 1 / 3.0  // KindFloat = KindFloat(1) / KindFloat(3.0)
 
-The third constant will be of kind float and represent 1/3 at a precision of 256 bits.
+A terceira constante será do tipo float e representará 1/3 com uma precisão de 256 bits.
 
     const zero = 1 / 3  // KindInt = KindInt(1) / KindInt(3)
 
-The zero constant will be of kind integer and set to 0 since integer division has
-no remainder.
+A constante "zero" será do tipo inteiro e definida como 0, pois a divisão de inteiros não tem resto.
```
Comment (rschio) [inline _content/tour/por/constants.article:57]:
```suggestion
A constante "zero" será do "kind" inteiro e definida como 0, pois a divisão de inteiros não tem resto.
```

**Review by rschio — APPROVED** [review]

**Pull request merged**